electron_app/src/ipc/backup.js

```diff
@@ -178,60 +214,134 @@ ipc.answerRenderer('restore-backup-encrypted', async params => {
+    if (acc1.triggerTimer > acc2.triggerTimer) return 1;
+    return 0;
+  });
+
+  if (nextBackupTimer) clearTimeout(nextBackupTimer);
+  nextBackupTimer = setTimeout(() => {
+    initAutoBackup(autoBackupsTime[0].accountId);
+  }, autoBackupsTime[0].triggerTimer);
 };
 
-const startBackupMonitor = account => {
-  const accountId = account.id;
-  clearTimeout(autoBackupIntervalId[accountId]);
+const initAutoBackup = async accountId => {
+  autoBackupsTime = autoBackupsTime.filter(timer => {
+    return timer.accountId !== accountId;
+  });
+  currentAutobackup = accountId;
+
+  const account = myAccount.loggedAccounts.find(acc => acc.id === accountId);
+  if (!account) {
+    backupDone();
+    return;
+  }
+
   const {
     autoBackupEnable,
     autoBackupPath,
     autoBackupFrequency,
     autoBackupNextDate
   } = account;
   if (!autoBackupEnable || !autoBackupNextDate) {
+    backupDone();
     return;
   }
-  const now = moment();
-  const pendingDate = moment(autoBackupNextDate);
-  const timeDiff = pendingDate.diff(now);
-  autoBackupIntervalId[accountId] = setTimeout(async () => {
-    try {
-      const backupFileName = defineBackupFileName('db');
-      const backupSize = await doExportBackupUnencrypted({
-        backupPath: `${autoBackupPath}/${backupFileName}`
-      });
-      const timeUnit = defineUnitToAppend(autoBackupFrequency);
-      const today = moment(Date.now());
-      const nextDate = moment(autoBackupNextDate);
-      do {
-        nextDate.add(1, timeUnit);
-      } while (nextDate.isBefore(today));
-      await updateAccount({
-        id: accountId,
-        autoBackupLastDate: pendingDate.format(backupDateFormat),
-        autoBackupLastSize: backupSize,
-        autoBackupNextDate: nextDate.format(backupDateFormat)
-      });
-      startBackupMonitor(account);
-    } catch (backupErr) {
-      log(
-        `Failed to do scheduled backup at ${pendingDate.format(
-          backupDateFormat
-        )}`
-      );
-    }
-  }, timeDiff);
+
+  try {
+    const backupFileName = defineBackupFileName('db');
+    const backupSize = await doExportBackupUnencrypted({
+      backupPath: `${autoBackupPath}/${backupFileName}`,
+      progressCallback: data => {
+        send('backup-progress', data);
+      }
+    });
+    const timeUnit = defineUnitToAppend(autoBackupFrequency);
+    const today = moment(Date.now());
+    const nextDate = moment(autoBackupNextDate);
+    do {
+      nextDate.add(1, timeUnit);
+    } while (nextDate.isBefore(today));
+    await updateAccount({
+      id: accountId,
+      autoBackupLastDate: today.format(backupDateFormat),
+      autoBackupLastSize: backupSize,
+      autoBackupNextDate: nextDate.format(backupDateFormat)
+    });
+    const timeDiff = nextDate.diff(today);
+    autoBackupsTime.push({
+      username: account.recipientId,
+      accountId,
+      triggerTimer: timeDiff <= 0 ? 1 : timeDiff
+    });
+
+    backupDone();
+  } catch (backupErr) {
+    log(
```

Comment:
same comment

Reply:
This log function wraps all console logs in the file, I even use it in PR #1467